### PR TITLE
Modified interface of ColumnProps<T> generic can check Error when a generic table

### PR DESCRIPTION
**Pull request by ghost**
fixed before can't check error 
image

fixed after
image

Can check error

image

First of all, thank you for your contribution! :-)

Please makes sure that these checkboxes are checked before submitting your pull request, thank you!

* [x] Make sure that you propose pull request to right branch: bugfix for `master`, feature for branch `feature`.
* [x] Make sure that you follow antd's [code convention](https://github.com/ant-design/ant-design/wiki/Code-convention-for-antd).
* [x] Run `npm run lint` and fix those errors before submitting in order to keep consistent code style.
* [x] Rebase before creating a pull request to keep commit history clear.
* [x] Add some descriptions and refer relative issues for you pull request.

Extra checklist:

**if** *isBugFix* **:**

  * [ ] Make sure that you add at least one unit test for the bug which you had fixed.

**elif** *isNewFeature* **:**

  * [ ] Update API docs for the component.
  * [x] Update/Add demo to demonstrate new feature.
  * [x] Update TypeScript definition for the component.
  * [ ] Add unit tests for the feature.

### CODE REVIEW

**Comment by yesmeck** [comment]
这个还是要用 `string`，`dataIndex` 不仅仅支持 record 的 key，还要支持嵌套的 key。

File: components/table/interface.tsx [hunk 1]
```diff
@@ -19,7 +19,7 @@ export interface ColumnProps<T> {
         },
       ) => React.ReactNode);
   key?: React.Key;
-  dataIndex?: string;
+  dataIndex?: keyof T;
```
Comment (yesmeck) [inline components/table/interface.tsx]: We can not use `ketof T` here, `dataIndex` not only accepts keys of record, also accepts the nested keys: `dataIndex: 'foo.bar'`.

Comment (yesmeck) [reply components/table/interface.tsx]: https://github.com/ant-design/ant-design/issues/9393

Comment (bondBo) [reply components/table/interface.tsx]: DataIndex 本身是table中最重要的取值key，模糊化失去原本检查的意义，如果使用嵌套key，可以手动去转型 or ColumnProps<T, U = string>, dataIndex : keyof U

**Pull request merged**